pallets/asset-manager/src/migrations.rs

```diff
@@ -0,0 +1,425 @@
+// Copyright 2019-2021 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{AssetIdType, AssetTypeId, AssetTypeUnitsPerSecond, Config};
+use frame_support::{
+	pallet_prelude::PhantomData,
+	storage::migration::storage_key_iter,
+	traits::{Get, OnRuntimeUpgrade},
+	weights::Weight,
+	Blake2_128Concat,
+};
+use sp_std::convert::TryInto;
+//TODO sometimes this is unused, sometimes its necessary
+use sp_std::vec::Vec;
+use xcm::latest::prelude::*;
+
+/// Migration that changes the mapping AssetId -> units_per_second to
+/// a mapping of the form AssetType -> units_per_second
+/// It does so by removing the AssetTypeUnitsPerSecond storage and
+/// populating the new AssetTypeUnitsPerSecond
+pub struct UnitsWithAssetType<T>(PhantomData<T>);
+impl<T: Config> OnRuntimeUpgrade for UnitsWithAssetType<T> {
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<(), &'static str> {
+		use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+
+		let pallet_prefix: &[u8] = b"AssetManager";
+		let storage_item_prefix: &[u8] = b"AssetIdUnitsPerSecond";
+
+		// We want to test that:
+		// There are no entries in the new storage beforehand
+		// The same number of mappings exist before and after
+		// As long as there are some mappings stored, one representative key maps to the
+		// same value after the migration.
+		// There are no entries in the old storage afterward
+
+		// Assert new storage is empty
+		assert!(AssetTypeUnitsPerSecond::<T>::iter().next().is_none());
```

Comment:
Out of curiosity what happens if this is failing? Will it break the chain ?

Reply:
this is only in pre_upgrade. So this only applies for the try-runtime tests. This does not get executed on_runtime()